src/dragonfs.c

```diff
@@ -684,36 +703,93 @@ int dfs_dir_findnext(char *buf)
     return __dfs_findnext(buf, &next_entry);
 }
 
+static uint32_t prime_hash(const char *str, uint32_t prime)
+{
+    uint32_t hash = 0;
+    while(*str) {
+        char c = *str++;
+        hash = (hash * prime) + c;
+    }
+    return hash;
+}
 
-int dfs_open(const char * const path)
+static int32_t lookup_file_hash_entry(uint32_t hash)
 {
-    /* Try to find file */
-    directory_entry_t *dirent;
-    int ret = recurse_path(path, WALK_OPEN, &dirent, TYPE_FILE);
+    int32_t left = 0;
+    int32_t right = file_lookup->num_files-1;
+    int32_t result = -1;
+    while(left <= right) {
+        int32_t mid = left+((right-left)/2);
+        if(file_lookup->files[mid].path_hash == hash) {
+            result = mid;
+            right = mid-1;
+        } else if(file_lookup->files[mid].path_hash < hash) {
+            left = mid+1;
+        } else {
+            right = mid-1;
+        }
+    }
+    return result;
+}
 
-    if(ret != DFS_ESUCCESS)
-    {
-        /* File not found, or other error */
-        return ret;
+static dfs_file_entry_t *lookup_file_path_entry(const char *const path, int32_t start_index, uint32_t hash)
+{
+    size_t buf_len = ROUND_UP(file_lookup->path_buf_size, 16);
+    char alignas(16) path_buf[buf_len];
```

Comment:
It seems that there is quite a bit of code to calculate this `path_buf_size` in mkdfs (it is the maximum path length as far as I can see), and that's just to make sure here to declare a stack buffer big enough.

Can't you just declare this buffer later, in the for cycle, just before the dma call? There, you can declare it exactly of `path_len` size, which is the length you need to read it. At each iteration of the loop, the length will change, but the compiler will allocate and release the stack allocation dynamically for you.

So I don't think that calculating this maximum size and keeping it stored in the lookup table is necessary.